synapse/handlers/oidc_handler.py

```diff
@@ -693,14 +697,33 @@ async def handle_redirect_request(
                 ui_auth_session_id=ui_auth_session_id,
             ),
         )
-        request.addCookie(
-            SESSION_COOKIE_NAME,
-            cookie,
-            path="/_synapse/client/oidc",
-            max_age="3600",
-            httpOnly=True,
-            sameSite="lax",
-        )
+
+        # we want the cookie to be returned to us even when the request is the POSTed
+        # result of a form on another domain, as is used with `response_mode=form_post`.
+        #
+        # Modern browsers will not do so unless we set SameSite=None; however *older*
+        # browsers (including all versions of Safari on iOS 12?) don't support
+        # SameSite=None, and interpret it as SameSite=Strict:
+        # https://bugs.webkit.org/show_bug.cgi?id=198181
+        #
+        # As a rather painful workaround, we set *two* cookies, one with SameSite=None
+        # and one with no SameSite, in the hope that at least one of them will get
+        # back to us.
+        #
+        # Secure is necessary for SameSite=None (and, empirically, also breaks things
+        # on iOS 12.)
+        #
+        # we have to build the cookie by hand rather than calling request.addCookie
+        # to work around https://twistedmatrix.com/trac/ticket/10088
+        #
+        for cookie_name, options in [
+            (SESSION_COOKIE_NAME, b"; Secure; SameSite=None"),
+            (SESSION_COOKIE_NAME_IOS_HACK, b""),
```

Comment:
These don't seem to match above where we clear out the cookies. In that case *both* are set to `Secure`.

We should probably pull these out to constants.

Reply:
Ooops. fair.